### PR TITLE
[JENKINS-70103] `WeakHashMap` broken for looking up Jetty `Session` in WebSocket response

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-70103](https://issues.jenkins.io/browse/JENKINS-70103). Amends #6780. Supersedes #7406.

### Testing done

Used `JNLPLauncherRealTest`.

### Proposed changelog entries

- Memory leak when repeatedly connecting WebSocket agents.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7412"></a>

### CODE REVIEW

File: core/src/main/java/jenkins/websocket/WebSockets.java
Comment (basil) [inline core/src/main/java/jenkins/websocket/WebSockets.java:71]: Visibility could be decreased:

```suggestion
                    private Object providerSession;
```

Generally desirable to keep visibility as low as possible to make future changes easier.

File: websocket/spi/src/main/java/jenkins/websocket/Provider.java
Comment (basil) [inline websocket/spi/src/main/java/jenkins/websocket/Provider.java:50]: Spelling:

```suggestion
        void onWebSocketConnect(Object providerSession);
```

**Review by basil — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/jenkins/websocket/WebSockets.java
Comment (jglick) [reply core/src/main/java/jenkins/websocket/WebSockets.java]: Not necessary in this case because it is an anonymous inner class.

**Pull request merged**
